module/state_synchronization/getter.go

```diff
@@ -0,0 +1,313 @@
+package state_synchronization
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/rs/zerolog"
+	"go.uber.org/atomic"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/blobs"
+	"github.com/onflow/flow-go/network"
+)
+
+const defaultMaxBlobTreeDepth = 1 // prevents malicious CID from causing download of unbounded amounts of data
+
+var ErrBlobTreeDepthExceeded = errors.New("blob tree depth exceeded")
+
+type ExecutionDataGetterOption func(*executionDataGetterImpl)
+
+func WithMaxBlobSize(size int) ExecutionDataGetterOption {
+	return func(s *executionDataGetterImpl) {
+		s.maxBlobSize = size
+	}
+}
+
+func WithMaxBlobTreeDepth(depth int) ExecutionDataGetterOption {
+	return func(s *executionDataGetterImpl) {
+		s.maxBlobTreeDepth = depth
+	}
+}
+
+func NewExecutionDataGetter(
+	codec encoding.Codec,
+	compressor network.Compressor,
+	blobService network.BlobService,
+	metrics module.ExecutionDataGetterMetrics,
+	logger zerolog.Logger,
+	opts ...ExecutionDataGetterOption,
+) *executionDataGetterImpl {
+	eds := &executionDataGetterImpl{
+		&serializer{codec, compressor},
+		blobService,
+		DefaultMaxBlobSize,
+		defaultMaxBlobTreeDepth,
+		metrics,
+		logger.With().Str("component", "execution_data_service").Logger(),
+	}
+
+	for _, opt := range opts {
+		opt(eds)
+	}
+
+	return eds
+}
+
+var _ ExecutionDataGetter = (*executionDataGetterImpl)(nil)
+
+type executionDataGetterImpl struct {
+	serializer       *serializer
+	blobService      network.BlobService
+	maxBlobSize      int
+	maxBlobTreeDepth int
+	metrics          module.ExecutionDataGetterMetrics
+	logger           zerolog.Logger
+}
+
+// retrieveBlobs retrieves the blobs for the given CIDs from the blobservice, and sends them to the given BlobSender
+// in the order specified by the CIDs.
+func (s *executionDataGetterImpl) retrieveBlobs(parent context.Context, bs *blobs.BlobSender, cids []cid.Cid, logger zerolog.Logger) error {
+	defer bs.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	blobChan := s.blobService.GetBlobs(ctx, cids) // initiate a batch request for the given CIDs
+	cachedBlobs := make(map[cid.Cid]blobs.Blob)
+	cidCounts := make(map[cid.Cid]int) // used to account for duplicate CIDs
+
+	for _, c := range cids {
+		cidCounts[c] += 1
+	}
+
+	for _, c := range cids {
+		blob, ok := cachedBlobs[c]
+
+		if !ok {
+			var err error
+
+			if blob, err = s.findBlob(ctx, blobChan, c, cachedBlobs, logger); err != nil {
+				return err
+			}
+		}
+
+		cidCounts[c] -= 1
+
+		if cidCounts[c] == 0 {
+			delete(cachedBlobs, c)
+		}
```

Comment:
Is this to be able to free memory taken by a cached blob earlier in the for loop?

Reply:
I don't quite remember tbh, it looks like that might have been the reason.